daemon/containerd/image_push.go

```diff
@@ -173,20 +173,25 @@ func (i *ImageService) pushRef(ctx context.Context, targetRef reference.Named, p
 		if cerrdefs.IsNotFound(err) && platform == nil {
 			// If push failed because of a missing content, no specific platform was requested
 			// and the target is an index, select a platform-specific manifest to push instead.
-			target, err = i.getPushDescriptor(ctx, img, nil)
+			var newTarget ocispec.Descriptor
+			newTarget, err = i.getPushDescriptor(ctx, img, nil)
 			if err != nil {
 				return err
 			}
 
-			pp.TurnNotStartedIntoUnavailable()
-			err = remotes.PushContent(ctx, pusher, target, store, limiter, platforms.All, handlerWrapper)
-
-			if err == nil {
-				progress.Aux(out, map[string]string{
-					"Stripped": "true",
-					"Index":    img.Target.Digest.String(),
-					"Manifest": target.Digest.String(),
-				})
+			// Retry only if the new push candidate is different from the previous one.
+			if newTarget.Digest != target.Digest {
+				target = newTarget
+				pp.TurnNotStartedIntoUnavailable()
+				err = remotes.PushContent(ctx, pusher, target, store, limiter, platforms.All, handlerWrapper)
+
+				if err == nil {
```

Comment:
Oh, right, so maybe there is something special with `err` here; because it's checked further down?

We should look if we can make this code easier to read; it's getting complex with quite some footguns.


Maybe this error (if it's indeed checked outside the branch) needs a special name to make it stand out 🤔 (or ideally we could make the code flow easier)

Reply:
Deduplicated some error handling so it's a bit cleaner. Still not perfect, but not sure how to make it better for now 😢

Reply:
Yeah, it's not a blocker, as it was already complex; definitely something we can handle in follow-ups. It was just that I tripped over the `err` .. and was confused, and that's a good indicator "this code may need some restructuring" 😄